src/Engine/Objects/ItemTable.cpp

```diff
@@ -625,39 +608,39 @@ void ItemTable::GenerateItem(ITEM_TREASURE_LEVEL treasure_level, unsigned int uT
         case EQUIP_CLOAK:
         case EQUIP_GAUNTLETS:
         case EQUIP_BOOTS:
-        case EQUIP_RING:
-
+        case EQUIP_RING: {
             if (!uBonusChanceStandart[treasure_level]) return;
-            special_chance = grng->random(100);
-            if (special_chance < uBonusChanceStandart[treasure_level]) {
-                v26 = grng->random(pEnchantmentsSumm[out_item->GetItemEquipType()]) + 1;
-                v27 = 0;
-                while (v27 < v26) {
-                    ++out_item->uEnchantmentType;
-                    v27 += pEnchantments[out_item->uEnchantmentType].to_item[out_item->GetItemEquipType()];
+            int bonusChanceRoll = grng->random(100);  // edx@86
+            if (bonusChanceRoll < uBonusChanceStandart[treasure_level]) {
+                int enchantmentChanceSumRoll = grng->random(chanceByItemTypeSums[outItem->GetItemEquipType()]) + 1;
+                int currentEnchantmentChancesSum = 0;
+                while (currentEnchantmentChancesSum < enchantmentChanceSumRoll) {
+                    currentEnchantmentChancesSum +=
+                        standardEnchantments[outItem->uEnchantmentType].chancesByItemType[outItem->GetItemEquipType()];
+                    ++outItem->uEnchantmentType;
                 }
 
-                v33 = grng->random(bonus_ranges[treasure_level].maxR - bonus_ranges[treasure_level].minR + 1);
-                out_item->m_enchantmentStrength =
-                        v33 + bonus_ranges[treasure_level].minR;
-                v32 = out_item->uEnchantmentType - 1;
-                if (v32 == 21 || v32 == 22 ||
-                    v32 == 23)  // Armsmaster skill, Dodge skill, Unarmed skill
-                    out_item->m_enchantmentStrength =
-                            out_item->m_enchantmentStrength / 2;
-                if (out_item->m_enchantmentStrength <= 0)
-                    out_item->m_enchantmentStrength = 1;
+                outItem->m_enchantmentStrength = bonusRanges[treasure_level].minR +
+                                                 grng->random(bonusRanges[treasure_level].maxR - bonusRanges[treasure_level].minR + 1);
+                PLAYER_SKILL_TYPE standardEnchantmentSkill = static_cast<PLAYER_SKILL_TYPE>(outItem->uEnchantmentType - 1);
+                if (standardEnchantmentSkill == PLAYER_SKILL_ARMSMASTER || standardEnchantmentSkill == PLAYER_SKILL_DODGE ||
+                    standardEnchantmentSkill == PLAYER_SKILL_UNARMED) {
+                    outItem->m_enchantmentStrength = outItem->m_enchantmentStrength / 2;
+                }
```

Comment:
This is not a valid change.

21 is "of Arms",
22 is "of Dodging",
23 is "of the Fist",

And skill enum has different numeric values.